### PR TITLE
27037-6-dynamic-endpoint-handlers

### PR DESCRIPTION
**Pull request by hdiedrich**
wip - simple build call; priv/ configs; no epi;
      only url_handlers are dynamic, the others,
      db_url_handlers/0 and design_url_handlers/0
      remain untouched in this commit;
      debug messages to screen;
      commented out test snippets in source;
      testing calls and exceptions for debugging robustnes

Config files of relevant modules are in their
27037-6-dynamic-endpoint-handlers branches.

Affected modules are:
      chttpd
      couch
      mem3
      global_changes
      setup
      mango

### CODE REVIEW

File: src/chttpd.erl [hunk 1]
```diff
@@ -111,6 +111,8 @@ start_link(Name, Options) ->
     ServerOptsCfg = config:get("chttpd", "server_options", "[]"),
     {ok, ServerOpts} = couch_util:parse_term(ServerOptsCfg),
     Options2 = lists:keymerge(1, lists:sort(Options1), lists:sort(ServerOpts)),
+    chttpd_handler:build(),
+    chttpd_handler:url_handler("x"),
```
Comment (kxepal) [inline src/chttpd.erl]: What is "x"?

File: src/chttpd_handler.erl [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+
+-spec build() -> ok | [].
+%% @doc Create the dynamic handler functions from ini file.
+build() ->
+    build(load_defs()).
+
+-spec build(HandlerCfg::config()) -> ok.
+%% @doc Compile the complete syntax tree, purge and load the dynamic module
+build(Cfg) when is_list(Cfg) ->
+    io:format(user, "~n➽ Dynamic handlers using config: ~n~p~n", [Cfg]),
```
Comment (kxepal) [inline src/chttpd_handler.erl]: Please use `couch_log` instead of `io:format/2`.

File: src/chttpd_handler.erl [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+    {db_url_handlers, list, lists:flatten(AllDBHandlers)},
+    {design_url_handlers, list, lists:flatten(AllDesignHandlers)}].
+
+-spec load_defs(AppName::atom()) -> OneAppsHandlerCfg::config().
+%% @doc assemble the configuration from the chttpd_handler.cfg of all apps.
+% load_defs(App) ->
+%    [ check_def(test_cfg(), "no path (testing)") ].
+load_defs(App) ->
+io:format(user, "~n➽ Dynamic handler uses priv dir for ~p: ~p~n", [App, code:priv_dir(App)]),    
+case code:priv_dir(App) of
```
Comment (kxepal) [inline src/chttpd_handler.erl]: Seems like indention is broken.

File: priv/chttpd_handler.cfg [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    {"_active_tasks",         {chttpd_misc, handle_task_status_req, 1}},
+    {"_node",                 {chttpd_misc, handle_node_req, 1}},
+    {"_reload_query_servers", {chttpd_misc,
+        handle_reload_query_servers_req, 1}},
+    {"_replicate",            {chttpd_misc, handle_replicate_req, 1}},
+    {"_uuids",                {chttpd_misc, handle_uuids_req, 1}},
+    {"_session",              {chttpd_auth, handle_session_req, 1}},
+    {"_up",                   {chttpd_misc, handle_up_req, 1}},
+    {"_user",                 {chttpd_auth, handle_user_req, 1}},
+    {'_',                     {chttpd_db,   handle_request, 1}}
```
Comment (kxepal) [inline priv/chttpd_handler.cfg]: Seems like "_oauth" accidentally missed, Can we also keep these lists sorted by resource name?

File: priv/chttpd_handler.cfg [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+%%
+%% Unless  required  by  applicable  law  or  agreed  to  in  writing,  software
+%% distributed  under the  License  is distributed on an "AS IS" BASIS,  WITHOUT
+%% WARRANTIES  OR  CONDITIONS  OF  ANY  KIND, either express or implied. See the
+%% License for the specific language governing permissions and limitations under
+%% the License.
+
+%% @doc link endpoints to callbacks - configs of multiple deps are combined.
+
+{docs, clauses, [
```
Comment (kxepal) [inline priv/chttpd_handler.cfg]: Why clauses, not list?

**Comment by kxepal** [comment]
This PR as like as https://github.com/apache/couchdb-chttpd/pull/10 missed support of configuring chttpd handlers via couch_config. Is that intentional?

File: priv/chttpd_handler.cfg (hunk 5 above)
Comment (kxepal) [inline priv/chttpd_handler.cfg]: As far as I understood, first element of this config structure defined to what API level following resources are belong: db for /db/:resource, design for /db/_design/ddoc/:resource etc. But why server root level is named as docs then? This causes false assumption that these resources will be available at /db/docid/ level.

File: priv/chttpd_handler.cfg [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    {"_all_dbs",              {chttpd_misc, handle_all_dbs_req, 1}},
+    {"_active_tasks",         {chttpd_misc, handle_task_status_req, 1}},
+    {"_node",                 {chttpd_misc, handle_node_req, 1}},
+    {"_reload_query_servers", {chttpd_misc,
+        handle_reload_query_servers_req, 1}},
+    {"_replicate",            {chttpd_misc, handle_replicate_req, 1}},
+    {"_uuids",                {chttpd_misc, handle_uuids_req, 1}},
+    {"_session",              {chttpd_auth, handle_session_req, 1}},
+    {"_up",                   {chttpd_misc, handle_up_req, 1}},
+    {"_user",                 {chttpd_auth, handle_user_req, 1}},
```
Comment (kxepal) [inline priv/chttpd_handler.cfg]: There is no such endpoint on current master nor such function in chttpd_auth module.

**Comment by hdiedrich** [comment]
Wrong repo

**Pull request closed without merging**